distributed/worker.py

```diff
@@ -1477,21 +1477,66 @@ def update_data(self, comm=None, data=None, report=True, serializers=None):
                 self.put_key_in_memory(ts, value)
                 ts.priority = None
                 ts.duration = None
+            ts.scheduler_holds_ref = True
 
             self.log.append((key, "receive-from-scatter"))
 
         if report:
+
+            self.log.append(
+                ("Notifying scheduler about in-memory in update-data", list(data))
+            )
             self.batched_stream.send({"op": "add-keys", "keys": list(data)})
         info = {"nbytes": {k: sizeof(v) for k, v in data.items()}, "status": "OK"}
         return info
 
-    def delete_data(self, comm=None, keys=None, report=True):
-        if keys:
-            for key in list(keys):
-                self.log.append((key, "delete"))
-                self.release_key(key, cause="delete data")
+    def free_keys_comm(self, comm, keys, reason):
+        """This is a comm handler calling `free_keys`"""
+        return self.handle_free_keys(keys, reason)
```

Comment:
Minor nit, but maybe we can remote this method, and add a `comm=None` keyword to the `handle_free_keys` method?  It would be unfortunate if we had to have two methods for every handler.

Reply:
At first, I had this in for debugging purposes to track where the scheduler actually called this. That's now obsolete since I am providing reasons everywhere (if we want to keep that).

Apart from this I liked this refactoring since it didn't require me to make all arguments optional which is nice for type checking. however, no strong opinions here.

On a deeper level, I'm wondering if we should allow the same functionality via stream and comm handlers. I'm inclined to say no since the semantics should then also be slightly different. For instance, comm handlers are useful if you actually require a response, otherwise a stream handler is preferable, for performance. Both is kind of weird. Thoughts?

Reply:
To be clear, the functionality you're talking about here is access to the `comm`?  If so, I guess the question reduces to "is there ever a time when we need the comm in either case?"  It's a valid question.  I don't have a strong thought here.

Reply:
yes. some handlers are merely signals which don't require a response. for instance "hey, scheduler. just wanted to let you know that I finished computing X. Cheers." 
Signals without a need for a response, i.e. handlers which do not require a comm, should be implemented as stream handlers since there we can batch submit all messages.

Anyhow, we should not mix this into the PR. I think I will simply remove the second method :)